### ISSUE TITLE
Kube-addons should be upgradeable

### ISSUE THREAD

**Issue by erictune** [GoogleCloudPlatform/kubernetes#7046]
If config in cluster/addons changes, then I'd like to be able to upgrade my cluster to pick it up.

The folks designing upgrades should decide how this is accomplished.  Probably one of:
  1. Do a kube-push, salt runs, kube-addons runs again, and it patches the existing configs.
  1. A specific upgrade script runs.

If 1 is the plan, then several things are broken at the moment.
 - When you try to run `salt-call state.highstate` on the master after modifying an addon, you get an error like: `:Failed when setting service kube-addons to start at boot, but the service is running`
 - Kube-addons forks children that run forever, and stopping the init script doesn't top the children
 - Kube-addons uses kubectl in a way that creates objects but won't update them.  It needs to either patch the objects or create-and-destroy them.

**Comment by zmerlynn** [comment]
Yeah, this is going to need work for v1.0. It actually runs infinitely to prevent you from running again, because salt was doing the wrong thing for a while, but the object issue is the much larger one. If the add-ons are now in their own namespace, I think we should just destroy the namespace and start over.

**Comment by dchen1107** [comment]
I moved this to 1.0 candidate because we are moving addon services to their own namespace now (#10331). 

cc @brendandburns @bgrant0607 @satnam6502

**Comment by dchen1107** [comment]
unassign myself from this one since I am overloaded already.

**Comment by alex-mohr** [comment]
Clusters need to be upgradable -- create at version 0.20.1, upgrade to 0.21.x, and they still work.  If addons were created using 0.20.1 and need to be changed in some way in order to work in 0.21,, then the upgrade-to-0.21 process must convert them.
Looks like current discussion is on #10487

**Comment by zmerlynn** [comment]
I thought this was covered already in #9128, did I miss something?

**Comment by alex-mohr** [comment]
#10487 is moving addons from one namespace to another.  The question is whether #9128 handles that case.

**Comment by zmerlynn** [comment]
@marekbiskup - If you get a chance to look at this soon, I'd very much
appreciate your expertise here.

We do have an e2e test for this, but I don't know if it's running
consistently). It should certainly be run on #10487, and I'll comment there.

**Comment by marekbiskup** [comment]
#9128 assumes the addons are always in the default namespace and does not work if namespaces change.

I'll take this issue. Probably I'll split it into several PRs.

 1. kill children of kube-addons.sh
 1. suppress salt error message `Failed when setting service kube-addons to start at boot`
 1. handle namespace changes.

**Comment by marekbiskup** [comment]
Question:

I want to find object with a specific label across **all** namespaces. Is there a way to do it?
If not, how can I locate the old add-ons (in an unknown namespace) to delete them?

**Comment by marekbiskup** [comment]
Ok, I see what @brendandburns / @satnam6502 did in #10487 was just to assume that previously all add-ons were in "default" namespace. It looks like namespace change is already handled in #10487.
I'll focus on other issues first.

**Comment by roberthbailey** [comment]
If we assume that addons are in the default namespace, is there further work to do here? Or is it only if we move addons to a different namespace?

**Comment by dchen1107** [comment]
cc/ @bgrant0607 @thockin and @davidopp since we just talked about moving addons to different namespace. If we think that one is p1, should we promote this one to p1 too?

**Comment by bgrant0607** [comment]
@marekbiskup @roberthbailey Future upgrades will need to be able to upgrade components in the new namespace. Assuming either default or kube-system for previously running add-ons seems fine.

@dchen1107 Yes, they should have the same priority.

**Comment by roberthbailey** [comment]
@bgrant0607 -- I wasn't specific enough; what I meant to ask was whether there was any work remaining that blocks v1.0 if we leave the addons in the default namespace. 

I was under the impression that the current addons updater scripts worked for addons in the default namespace but not if we move them to a new namespace. In this case, if we wait to move them, then we could bump the remaining work here out of the 1.0 milestone as well.

**Comment by marekbiskup** [comment]
@roberthbailey 

There are two more things @erictune mentioned:
 1. kill children of kube-addons.sh (patch almost ready)
 1. suppress salt error message Failed when setting service kube-addons to start at boot (to be investigated)

**Comment by marekbiskup** [comment]
I don't see the salt error anymore. It must have been fixed at some point.
Addon update is already working.

#10611 only will resolve this issue.

**Comment by wojtek-t** [comment]
@marekbiskup - can you summarize what is left to do here (if anything?)

**Comment by marekbiskup** [comment]
See my comment just above.

**Comment by zmerlynn** [comment]
The last piece of this was merged. I believe this issue is done. Someone can reopen is this is incorrect!
